aten/src/ATen/core/dispatch/Dispatcher.h

```diff
@@ -399,48 +438,71 @@ inline Return Dispatcher::callWithDispatchKey(const TypedOperatorHandle<Return(A
 }
 
 template<class Return, class... Args>
 inline Return Dispatcher::redispatch(const TypedOperatorHandle<Return (Args...)>& op, DispatchKey currentDispatchKey, Args... args) const {
   detail::unused_arg_(args...);  // workaround for a false-positive warning about unused parameters in gcc 5
-  auto dispatchKey = op.operatorIterator_->op.dispatchKeyExtractor()
-    .template getDispatchKeyUnboxed<Args...>(
+  auto currentDispatchKeySet = op.operatorIterator_->op.dispatchKeyExtractor()
+    .template getDispatchKeySetUnboxed<Args...>(
       DispatchKeySet(DispatchKeySet::FULL_AFTER, currentDispatchKey),
       args...
     );
   // do not use RecordFunction on redispatch
-  const KernelFunction& kernel = op.operatorIterator_->op.lookup(dispatchKey);
-  return kernel.template call<Return, Args...>(op, std::forward<Args>(args)...);
+  const KernelFunction& kernel = op.operatorIterator_->op.lookup(currentDispatchKeySet.highestPriorityTypeId());
+  return kernel.template call<Return, Args...>(op, currentDispatchKeySet, std::forward<Args>(args)...);
+}
+
+template<class Return, class... Args>
+inline Return Dispatcher::callWithPrecomputedDispatchKeySet(const TypedOperatorHandle<Return (Args...)>& op, DispatchKeySet currentDispatchKeySet, Args... args) const {
+  detail::unused_arg_(args...);  // workaround for a false-positive warning about unused parameters in gcc 5
+  // do not use RecordFunction on redispatch
+  const KernelFunction& kernel = op.operatorIterator_->op.lookup(currentDispatchKeySet.highestPriorityTypeId());
+  return kernel.template call<Return, Args...>(op, currentDispatchKeySet, std::forward<Args>(args)...);
 }
 
 inline void Dispatcher::callBoxed(const OperatorHandle& op, Stack* stack) const {
   // note: this doesn't need the mutex because write operations on the list keep iterators intact.
   const auto& entry = op.operatorIterator_->op;
-  auto dispatchKey = entry.dispatchKeyExtractor().getDispatchKeyBoxed(stack);
-  const auto& kernel = entry.lookup(dispatchKey);
+  auto dispatchKeySet = entry.dispatchKeyExtractor().getDispatchKeySetBoxed(stack);
+  const auto& kernel = entry.lookup(dispatchKeySet.highestPriorityTypeId());
+  return _callBoxed(op, entry, kernel, dispatchKeySet, stack);
+}
+
+inline void Dispatcher::callBoxed(const OperatorHandle& op, DispatchKeySet dispatchKeySet, Stack* stack) const {
+  // note: this doesn't need the mutex because write operations on the list keep iterators intact.
+  const auto& entry = op.operatorIterator_->op;
+  const auto& kernel = entry.lookup(dispatchKeySet.highestPriorityTypeId());
+  return _callBoxed(op, entry, kernel, dispatchKeySet, stack);
+}
 
+// We always want to inline this helper, since it was originally part of callBoxed().
+C10_ALWAYS_INLINE void Dispatcher::_callBoxed(const OperatorHandle& op, const impl::OperatorEntry& entry, const KernelFunction& kernel, DispatchKeySet dispatchKeySet, Stack* stack) const {
```

Comment:
also here. The general guidance is to use always_inline very sparingly.
* The compiler usually knows better
* The compiler is flexible across architectures and can decide to inline on some and not on others
* The compiler is more flexible for future code changes. Even if always_inline is optimal today, it might be suboptimal next week after further code changes and you wouldn't notice but the compiler would.

So only use always_inline if it shows a significant difference in benchmarks to be worth those risks and potential degradations in the future and on other architectures..

Reply:
Thanks for the background- I was probably focusing too much on "don't cause a perf regression just so I can increase readability".

That point is interesting though- by previously hardcoding this function into its two call sites, that I was technically making a (minuscule) perf decision too, not just a code-quality one (i.e. assuming that hardcoding that function body into its two call-sites would also be better for perf).

Reply:
I just want to give a counterargument here, which I mentioned to @bdhirsh in our 1:1, which is that sometimes you really *do* know what assembly you're going for, and yes, no actually, I want all of this inlined, and applying always inline is going to be the most predictable way to get the assembly you want. The trick is to actually be reasonably confident in the assembly you want, and if you aren't actually looking at it it's easy to do something bad (like accidentally inline a big chunk of cold code). But for something like dispatcher, I'd argue that we ought to have a pretty good handle on what it should look like.

Reply:
Just fyi, I replaced `C10_ALWAYS_INLINE` with `inline` on those two functions and saw a pretty substantial regression in instruction counts (the first benchmark that I tested was `at::empty`).

I'm going to play around for a bit with moving everything RecordFunction-related into its own function, since everything outside of `shouldRunRecordFunction()` is hot and should probably be inlined and everything inside probably shouldn't. Failing that, I'm leaning towards retaining `C10_ALWAYS_INLINE` to avoid the regression.